File: scripts/ingest-docs-api.py
# Usage: python3 ingest_files.py /path/to/your/directory_or_file --ingest-secret YOUR_SECRET

import os
import requests
import argparse
import base64
from pathlib import Path

def is_valid_file(file_path):
    return os.path.isfile(file_path) and not os.path.basename(file_path).startswith('.')

def send_file_to_api(file_path, api_url, ingest_secret):
    with open(file_path, 'rb') as file:
        content = base64.b64encode(file.read()).decode('utf-8')
    
    metadata = {"source": os.path.basename(file_path)}
    
    payload = {
        "content": content,
        "metadata": metadata
    }
    
    headers = {
        "X-Ingest-Secret": ingest_secret
    }
    
    response = requests.post(api_url, json=payload, headers=headers)
    return response

def process_path(path, api_url, ingest_secret):
    if os.path.isfile(path):
        if is_valid_file(path):
            print(f"Sending file: {path}")
            response = send_file_to_api(path, api_url, ingest_secret)
            if response.status_code == 200:
                print(f"Successfully ingested: {path}")
            else:
                print(f"Failed to ingest {path}. Status code: {response.status_code}")
        else:
            print(f"Skipping invalid or hidden file: {path}")
    elif os.path.isdir(path):
        for root, _, files in os.walk(path):
            for file in files:
                file_path = os.path.join(root, file)
                process_path(file_path, api_url, ingest_secret)
    else:
        print(f"Error: {path} is not a valid file or directory.")

def main(path, api_url, ingest_secret):
    process_path(path, api_url, ingest_secret)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Ingest files from a directory or a single file to the API")
    parser.add_argument("path", help="Path to the directory or file to ingest")
    parser.add_argument("--endpoint", default="http://localhost:5328", help="API endpoint URL")
    parser.add_argument("--ingest-secret", required=True, help="Secret key for API authentication")
    args = parser.parse_args()

    main(args.path, f"{args.endpoint}/api/ingest", args.ingest_secret)

File: scripts/ingest-docs-api.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const { program } = require('commander');

function isValidFile(filePath) {
  return fs.statSync(filePath).isFile() && !path.basename(filePath).startsWith('.');
}

async function sendFileToApi(filePath, apiUrl, ingestSecret) {
  const content = fs.readFileSync(filePath, { encoding: 'base64' });
  const metadata = { 'source': path.basename(filePath) };
  const payload = { 'content': content, 'metadata': metadata };

  try {
    console.log(`Sending: ${filePath} to ${apiUrl}`);
    const response = await axios.post(apiUrl, payload, {
      headers: {
        'Content-Type': 'application/json',
        'X-Ingest-Secret': ingestSecret
      }
    });
    return response;
  } catch (error) {
    console.error(`Error sending file: ${error.message}`);
    return { status: error.response ? error.response.status : 500 };
  }
}

function isExcluded(filePath, excludeFiles) {
  return excludeFiles.some(pattern => {
    const regex = new RegExp('^' + pattern.replace(/\*/g, '.*').replace(/\?/g, '.') + '$');
    return regex.test(filePath);
  });
}

async function processFile(filePath, apiUrl, ingestSecret) {
  console.log(`Sending file: ${filePath}`);
  const response = await sendFileToApi(filePath, apiUrl, ingestSecret);
  if (response.status === 200) {
    console.log(`Successfully ingested: ${filePath}`);
  } else {
    console.log(`Failed to ingest ${filePath}. Status code: ${response.status}`);
  }
}

async function ingestFiles(directoryPath, apiUrl, ingestSecret, excludeFiles = []) {
  if (!fs.existsSync(directoryPath)) {
    console.error(`Error: ${directoryPath} does not exist.`);
    return;
  }

  const stats = fs.statSync(directoryPath);
  if (stats.isFile()) {
    if (isValidFile(directoryPath) && !isExcluded(directoryPath, excludeFiles)) {
      await processFile(directoryPath, apiUrl, ingestSecret);
    }
    return;
  }

  if (!stats.isDirectory()) {
    console.error(`Error: ${directoryPath} is not a valid directory or file.`);
    return;
  }

  const files = fs.readdirSync(directoryPath, { withFileTypes: true });

  for (const file of files) {
    const filePath = path.join(directoryPath, file.name);
    if (file.isDirectory()) {
      await ingestFiles(filePath, apiUrl, ingestSecret, excludeFiles);
    } else if (isValidFile(filePath) && !isExcluded(filePath, excludeFiles)) {
      await processFile(filePath, apiUrl, ingestSecret);
    } else {
      console.log(`Skipping invalid, hidden, or excluded file: ${filePath}`);
    }
  }
}

program
  .description('Ingest files from a directory to the API')
  .argument('<path>', 'Path to the directory containing files to ingest')
  .requiredOption('--ingest-secret <secret>', 'Ingest secret for API authentication')
  .option('--endpoint <url>', 'API endpoint URL', 'http://localhost:5328')
  .option('--exclude <files>', 'Comma-separated list of files or patterns to exclude', 'node_modules,.git,.env,package-lock.json')
  .action(async (directoryPath, options) => {
    const apiUrl = `${options.endpoint}/api/ingest`;
    const excludeFiles = options.exclude.split(',').map(item => item.trim());
    await ingestFiles(directoryPath, apiUrl, options.ingestSecret, excludeFiles);
  });

program.parse(process.argv);